lib/devtools/rake/flay.rb

```diff
@@ -0,0 +1,110 @@
+module Devtools
+  module Rake
+    # Flay metric runner
+    class Flay
+      include Anima.new(:threshold, :total_score, :lib_dirs, :excludes), Procto.call(:verify), Adamantium
+
+      BELOW_THRESHOLD = 'Adjust flay threshold down to %d'.freeze
+      TOTAL_MISMATCH  = 'Flay total is now %d, but expected %d'.freeze
+      ABOVE_THRESHOLD = '%d chunks have a duplicate mass > %d'.freeze
+
+      # Verify code specified by `files` does not violate flay expectations
+      #
+      # @raise [SystemExit] if a violation is found
+      # @return [undefined] otherwise
+      #
+      # @api private
+      def verify
+        # Run flay first to ensure the max mass matches the threshold
+        Devtools.notify_metric_violation(BELOW_THRESHOLD % largest_mass) if below_threshold?
+        Devtools.notify_metric_violation(TOTAL_MISMATCH % [total_mass, total_score]) if total_mismatch?
+
+        # Run flay a second time with the threshold set
+        return unless above_threshold?
+
+        restricted_flay_scale.flay_report
+        Devtools.notify_metric_violation(ABOVE_THRESHOLD % [restricted_mass_size, threshold])
+      end
+
+      private
+
+      def files
+        Devtools::Flay::FileList.call(lib_dirs, excludes).map(&:to_s)
+      end
+
+      # Is there mass duplication which exceeds specified `threshold`
+      #
+      # @return [Boolean]
+      #
+      # @api private
+      def above_threshold?
+        restricted_mass_size.nonzero?
+      end
+
+      # Is the specified `threshold` greater than the largest flay mass
+      #
+      # @return [Boolean]
+      #
+      # @api private
+      def below_threshold?
+        threshold > largest_mass
+      end
+
+      # Is the expected mass total different from the actual mass total
+      #
+      # @return [Boolean]
+      #
+      # @api private
+      def total_mismatch?
+        !total_mass.equal?(total_score)
+      end
+
+      # Size of mass measured by `Flay::Scale` and filtered by `threshold`
+      #
+      # @return [Integer]
+      #
+      # @api private
+      def restricted_mass_size
+        restricted_flay_scale.measure.size
+      end
+
+      # Sum of all flay mass
+      #
+      # @return [Integer]
+      #
+      # @api private
+      def total_mass
+        flay_masses.inject(:+).to_i
```

Comment:
> Is there a case where flay_masses is an empty array and so inject should get an explicit start accumulator?

Flay should always return an empty hash for `flay.masses` which means `Scale#measure` should always produce at least an empty array:

```ruby
flay = Flay.new(mass: 1_000)
flay.process(*Dir['lib/**/*.rb'])
flay.analyze
flay.masses # => {}
flay.masses.map { |hash, mass| Rational(mass, flay.hashes.fetch(hash).size) } # => []
```